server/src/main/java/org/elasticsearch/index/query/MatchNoneQueryBuilder.java

```diff
@@ -25,15 +25,28 @@
 public class MatchNoneQueryBuilder extends AbstractQueryBuilder<MatchNoneQueryBuilder> {
     public static final String NAME = "match_none";
 
+    private QueryBuilder originalQueryBuilder;
+
     public MatchNoneQueryBuilder() {}
 
+    public MatchNoneQueryBuilder(QueryBuilder originalQueryBuilder) {
```

Comment:
I think we should make this more generic by not only taking another query builder from rewrite as an argument. I was thinking more along the lines of taking a simple "reason" string or something we can than directly use in the newMatchNoDocsQuery() call. In most cases in the rewrite phase this can currently be just a string like the one used here (ie. The XYZ query was rewritten to a "match_none" query), but it cases like e.g. the range query rewrite we could also in the future include information why the rewrite happened (e.g. for range why the getRelation() call returned Relation.DISJOINT etc...) This doesn't have to happen in this PR, but I'd like to keep the option for future extensions.

Reply:
This make sense. I will make the change.